onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -51,8 +54,28 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       node->SetExecutionProviderType(kCpuExecutionProvider);
     }
 
+    // Check whether the CPU EP has a kernel for this node. Stop proceeding if it doesn't.
+    std::unique_ptr<CPUExecutionProvider> cpu_execution_provider =
+        onnxruntime::make_unique<CPUExecutionProvider>(CPUExecutionProviderInfo());
+
+    auto cpu_kernel_found_for_node =
+        KernelRegistry::HasImplementationOf(*cpu_execution_provider->GetKernelRegistry(), *node, kCpuExecutionProvider);
+
```

Comment:
Could we maybe restructure OptimizerExecutionFrame::Info::Info instead to separate out the kernel creation step so it can return an error so we don't need extra logic here?

Reply:
Then the call site requires an additional call to invoke kernel creation and check for the status and besides this way, it is more in tune with the theme in constant folding where we decide whether this node is a candidate for folding or not. Such a theme does not exist in OptimizerExecutionFrame::Info

Reply:
Isn't it more re-usable to do this inside OptimizerExecutionFrame::Info as its role (at least currently) involves creating the kernels for nodes it's given? Otherwise any user of that must first check if TryCreateKernel will fail prior to asking OptimizerExecutionFrame::Info to create the kernels.

That also avoids duplicating checks here that happen inside of TryCreateKernel.

Reply:
So do we expose a new method that returns a Status and pretty much does everything in Info's ctor ?

Reply:
Refactored